src/auth.zig

```diff
@@ -5,23 +5,137 @@ const std = @import("std");
 const Allocator = std.mem.Allocator;
 
 pub const Auth = @This();
+pub const User = @import("auth/user.zig");
 
-user: User = User{},
+provider: AnyAuth,
+current_user: ?User = null,
 
-pub const User = struct {
-    username: []const u8 = "invalid username",
+pub const Error = error{
+    UnknownUser,
+    Unauthenticated,
+    NotProvided,
 };
 
 pub fn valid(a: Auth) bool {
-    _ = a;
-    return true;
+    return a.provider.valid() catch false;
 }
 
-pub fn validOrError(a: Auth) !void {
-    if (!a.valid()) return error.Unauthenticated;
+pub fn requireValid(a: Auth) !void {
+    if (a.current_user == null or !a.valid()) return error.Unauthenticated;
 }
 
-pub fn currentUser(a: Auth, alloc: Allocator) !User {
-    _ = alloc;
-    return a.user;
+pub fn Provider(T: type) type {
+    return struct {
+        const Self = @This();
+        ctx: T,
+
+        pub fn init(ctx: T) Self {
+            return .{
+                .ctx = ctx,
+            };
+        }
+
+        /// TODO document the implications of non consttime function
+        pub fn lookupUser(self: *const Self, user_id: []const u8) Error!User {
```

Comment:
Rather than defining a User type in verse should we allow the caller to supply their own User?

Reply:
Yes, and it took a lot of self control to not do both in this one commit. There are a few things that need to exist (like a unique user identity), and I suspect we'll have something to wrap that. But I did think this stand in would be good enough for now? I can delete it and use a usize throughout (as the user_id) until we actually decide what it should be if you think it should start as a minimal abstraction?

Reply:
Keeping it for now sounds good.